File: LATIHAN_ARRAY/js/c9.js
function validateArrayData(arr) {
  // 1️⃣ Statistik
  const stats = {
    total: arr.length,
    numbers: arr.filter(x => typeof x === "number").length,
    strings: arr.filter(x => typeof x === "string").length,
    nulls: arr.filter(x => x === null).length,
    undefined: arr.filter(x => x === undefined).length
  };

  // 2️⃣ Validasi
  const allNumbers = arr.every(x => typeof x === "number");
  const hasNulls = arr.some(x => x === null);
  const hasUndefined = arr.some(x => x === undefined);
  const isValid = allNumbers && !hasNulls && !hasUndefined;

  // 3️⃣ Cleaned array (hanya angka)
  const cleaned = arr.filter(x => typeof x === "number");

  return { isValid, allNumbers, hasNulls, hasUndefined, stats, cleaned };
}

// ===== Driver Code =====
let data1 = [1, 2, 3, 4, 5];
let data2 = [1, "2", 3, null, 5, undefined];

console.log(validateArrayData(data1));
console.log(validateArrayData(data2));

File: projects/experimental/challenge/LATIHAN_ARRAY/rs/src/c_array/c9.rs
#[allow(dead_code)]
#[derive(Debug)]
// Enum ini digunakan untuk merepresentasikan tipe data yang berbeda-beda dalam satu array.
// compiler Rust akan memaksa kita memeriksa semua kemungkinan lewat match,
pub enum Value {
    Number(i32),
    Str(String),
    Null,
    Undefined,
}
#[allow(dead_code)]
#[derive(Debug)]
pub struct Stats {
    pub total: usize,
    pub numbers: usize,
    pub strings: usize,
    pub nulls: usize,
    pub undefined: usize,
}
#[allow(dead_code)]
#[derive(Debug)]
pub struct ArrayValidation {
    pub is_valid: bool,
    pub all_numbers: bool,
    pub has_nulls: bool,
    pub has_undefined: bool,
    pub stats: Stats,
    pub cleaned: Vec<i32>,
}

pub fn validate_array_data(arr: &Vec<Value>) -> ArrayValidation {
    let mut numbers_count = 0;
    let mut strings_count = 0;
    let mut nulls_count = 0;
    let mut undefined_count = 0;
    let mut cleaned = Vec::new();

    for item in arr.iter() {
        match item {
            // Pakai match untuk mendeteksi tipe Value
            // Jika Number, tambahkan ke cleaned
            Value::Number(n) => {
                numbers_count += 1;
                cleaned.push(*n);
            }
            Value::Str(_) => {
                strings_count += 1;
            }
            Value::Null => {
                nulls_count += 1;
            }
            Value::Undefined => {
                undefined_count += 1;
            }
        }
    }

    let total = arr.len();
    let all_numbers = numbers_count == total;
    let is_valid = all_numbers && nulls_count == 0 && undefined_count == 0;
    let has_nulls = nulls_count > 0;
    let has_undefined = undefined_count > 0;

    ArrayValidation {
        is_valid,
        all_numbers,
        has_nulls,
        has_undefined,
        stats: Stats {
            total,
            numbers: numbers_count,
            strings: strings_count,
            nulls: nulls_count,
            undefined: undefined_count,
        },
        cleaned,
    }
}
